vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml

```diff
@@ -496,40 +471,19 @@
     failed_when: upgraded_cp_nodes_count.stdout|int != 0
 
   - name: Get control plane machines
-    shell: | 
-            kubectl get machines -n "{{ NAMESPACE }}" -l  cluster.x-k8s.io/control-plane -o json |
-            jq -r '[ .items[] | select(.spec.version == "{{ UPGRADED_K8S_VERSION }}") | .status.nodeRef.name ]'
+    shell: |
+            kubectl get machines -n "{{ NAMESPACE }}" -l cluster.x-k8s.io/control-plane -o json 
     environment:
       KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
     register: new_control_plane_nodes
+    failed_when: >
+      (new_control_plane_nodes.stdout == "[]") or
+      (new_control_plane_nodes.stderr != "") or
+      (new_control_plane_nodes.rc > 1)
 
-  - name: Extract the list of new node names
+  - name: Extract the new node name
     set_fact:
-      new_control_plane_nodes: "{{ new_control_plane_nodes.stdout | from_json }}"
-
-  - name: Scale worker up to 1
-    kubernetes.core.k8s:
-      api_version: cluster.x-k8s.io/{{ CAPI_VERSION }}
-      kind: MachineDeployment
-      name: "{{ CLUSTER_NAME }}"
-      namespace: "{{ NAMESPACE }}"
-      resource_definition:
-        spec:
-          replicas: 1
-      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-
-  - name: Wait until worker has joined the cluster
-    kubernetes.core.k8s_info:
-      api_version: v1
-      kind: Node
-      label_selectors:
-        - "!node-role.kubernetes.io/control-plane"
-      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    retries: 200
-    delay: 20
-    register: worker_nodes
-    until: (worker_nodes is succeeded) and
-           (worker_nodes.resources | length == 1)
+      new_cp_node_name: "{{ new_control_plane_nodes.stdout | from_json | first }}"
 
   - name: Label worker for scheduling purpose
```

Comment:
Since you are removing `worker_nodes` in the previous task, this will be failing as it will try to access it.